apiserver/pkg/http/client.go

```diff
@@ -638,29 +649,92 @@ func (krc *KuberayAPIServerClient) doDelete(deleteURL string) (*rpcStatus.Status
 }
 
 func (krc *KuberayAPIServerClient) executeRequest(httpRequest *http.Request, URL string) ([]byte, *rpcStatus.Status, error) {
-	response, err := krc.httpClient.Do(httpRequest)
-	if err != nil {
-		return nil, nil, fmt.Errorf("failed to execute http request for url '%s': %w", URL, err)
-	}
-	defer func() {
-		if closeErr := response.Body.Close(); closeErr != nil {
-			klog.Errorf("Failed to close http response body because %+v", closeErr)
+	// Set the overall timeout
+	ctx, cancel := context.WithTimeout(context.Background(), krc.retryCfg.OverallTimeout)
+	defer cancel()
+
+	httpRequest = httpRequest.WithContext(ctx)
+
+	// Record the last error and status got
+	var lastErr error
+	var lastStatus *rpcStatus.Status
+
+	doReq := func() ([]byte, int, error) {
+		response, err := krc.httpClient.Do(httpRequest)
+		if err != nil {
+			return nil, 0, fmt.Errorf("failed to execute http request for url '%s': %w", URL, err)
 		}
-	}()
-	bodyBytes, err := io.ReadAll(response.Body)
-	if err != nil {
-		return nil, nil, fmt.Errorf("failed to read response body bytes: %w", err)
+		defer response.Body.Close()
+
+		bodyBytes, err := io.ReadAll(response.Body)
+		// Error in reading response body, treated as non-retryable error
+		if err != nil {
+			return nil, 0, fmt.Errorf("failed to read response body bytes: %w", err)
+		}
+
+		return bodyBytes, response.StatusCode, err
 	}
-	if response.StatusCode != http.StatusOK {
+
+	// Only retry for HTTP status codes defined as retryable in isRetryableHTTPStatus().
+	for attempt := 0; attempt <= krc.retryCfg.MaxRetry; attempt++ {
+		bodyBytes, statusCode, err := doReq()
+		// Error in sending the request, treated as non-retryable error
+		if err != nil {
+			lastStatus = nil
+			lastErr = err
+			break
+		}
+
+		if statusCode == http.StatusOK {
+			return bodyBytes, nil, nil
+		}
+
 		status, err := krc.extractStatus(bodyBytes)
+		// Error in extracting status from response body, treated as non-retryable error
 		if err != nil {
-			return nil, nil, err
+			lastStatus = nil
+			lastErr = err
+			break
 		}
-		return nil, status, &KuberayAPIServerClientError{
-			HTTPStatusCode: response.StatusCode,
+
+		lastStatus = status
+		lastErr = &KuberayAPIServerClientError{
+			HTTPStatusCode: statusCode,
+		}
+
+		// Retry only for HTTP status in the list
+		if !isRetryableHTTPStatus(statusCode) {
+			break
 		}
+
+		// Backoff before retry
+		sleep := krc.retryCfg.InitBackoff * time.Duration(math.Pow(krc.retryCfg.BackoffFactor, float64(attempt)))
+		if sleep > krc.retryCfg.MaxBackoff {
+			sleep = krc.retryCfg.MaxBackoff
+		}
+
+		select {
+		case <-time.After(sleep):
+			// continue to the next retry after backoff
+		case <-ctx.Done():
+			return nil, lastStatus, fmt.Errorf("overall timeout reached: %w", ctx.Err())
+		}
+
+	}
+	return nil, lastStatus, lastErr
+}
+
+func isRetryableHTTPStatus(statusCode int) bool {
+	switch statusCode {
```

Comment:
How about creating a `map[T]struct{}` for all retryable HTTP status codes and looking up the map? The `switch` statement and the utility function are not necessary in my opinion.

Reply:
Or even use an array.
Array outperforms hash map when item number is small (no need to calculate hash value, better cache locality)